### PR TITLE
SOF-251: Unit tests - Incomplete Sessions Screen

### PR DESCRIPTION
**Pull request by JohnnieShen**
This PR implements tests for the incomplete session screen, it covers all regular workflows as well as edge cases. Tests should be tested by running them on a physical device, both UI and VM tests.

### CODE REVIEW

File: app/src/main/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreen.kt [hunk 1]
```diff
@@ -37,18 +41,21 @@ fun IncompleteSessionScreen(
...
+                    }
+                    .testTag(IncompleteSessionTestTags.BACK_BUTTON) )
         },
         modifier = modifier
     ) {
         items(
             items = state.sessions.asReversed(),
             key = { it.localId }
         ) { session ->
+            val index = reversed.indexOf(session)
```
Comment (aryamanshodhan) [inline app/src/main/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreen.kt]: Suggestion [nitpick]: If you need to get access to the index, you can replace items(...) on line 49 to itemsIndexed(...). That way you can get rid of reversed on line 30.

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    private lateinit var currentSessionCache: CurrentSessionCache
+    private lateinit var cameraRepository: CameraRepository
+    private lateinit var viewModel: IncompleteSessionViewModel
+
+    private lateinit var incompleteSessionsFlow: MutableStateFlow<List<Session>>
+
+    @Before
+    fun setUp() {
+        mockkStatic(Uri::class)
+        mockkStatic(Log::class)
```
Comment (aryamanshodhan) [inline app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt:50]: Question: Why are we mocking Log?

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+        viewModel = IncompleteSessionViewModel(
+            sessionRepository = sessionRepository,
+            currentSessionCache = currentSessionCache,
+            cameraRepository = cameraRepository
+        )
+    }
+
+    @After
+    fun tearDown() {
+        unmockkStatic("android.net.Uri")
```
Comment (aryamanshodhan) [inline app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt:74]: Question: Can we not say unmockkStatic(Uri::class) and unmockkStatic(Log::class)?

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+    }
+
+    // ========================================
+    // Test Harness / Helpers
+    // ========================================
+
+    private fun makeSession(sessionType: SessionType) = Session(
+        localId = UUID.randomUUID(),
+        remoteId = null,
+        houseNumber = "1",
```
Comment (aryamanshodhan) [inline app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt]: Session does not have house number anymore

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+            advanceUntilIdle()
+            expectNoEvents()
+        }
+
+        coVerify(exactly = 1) { ErrorMessageBus.emit(IncompleteSessionError.SESSION_RETRIEVAL_FAILED, any()) }
+        coVerify(exactly = 0) { currentSessionCache.saveSession(any(), any()) }
+    }
+
+    @Test
+    fun incVm_b04_multipleResumes_emitMultipleNavigateEvents_andSaveEachTime() = runTest {
```
Comment (aryamanshodhan) [inline app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt:197]: Question: Would we encounter this situation in a real workflow?

File: app/src/androidTest/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreenTest.kt [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        val now = System.currentTimeMillis()
+        val day = 24L * 60 * 60 * 1000
+
+        fun session(
+            type: SessionType,
+            createdAt: Long
+        ) = Session(
+            localId = UUID.randomUUID(),
+            remoteId = null,
+            houseNumber = "",
```
Comment (aryamanshodhan) [inline app/src/androidTest/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreenTest.kt]: House number is no longer a part of session

File: app/src/androidTest/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreenTest.kt [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+        const val Intake = "intake/{type}"
+
+        fun intakeRouteFor(type: SessionType) = "intake/${type.name}"
+    }
+
+    private fun sampleSessions(): List<Session> {
+        val now = System.currentTimeMillis()
+        val day = 24L * 60 * 60 * 1000
+
+        fun session(
```
Comment (aryamanshodhan) [inline app/src/androidTest/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionScreenTest.kt]: Suggestion: Maybe rename this to makeSession like how you defined it in the IncompleteSessionViewModelTest

**Review by aryamanshodhan — CHANGES_REQUESTED** [review]
This is VERY comprehensive! I love it. Well done Johnny. There are a couple of changes needed to make it compatible with our new architecture. Minor changes needed.

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt (hunk 2 above)
Comment (JohnnieShen) [reply app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt:50]: There was an persistent error if you don't mock log, I don't quite remember what it was but mocking log is the only solution I figured out for it

File: app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt (hunk 5 above)
Comment (JohnnieShen) [reply app/src/test/java/com/vci/vectorcamapp/incomplete_session/presentation/IncompleteSessionViewModelTest.kt:197]: Probably not, unless you double tap on a slow phone (which i believe is still relevant), I can update this test to better reflect the behavior we want, e.g. when double tap, we don't really care about intermediate steps but as long as the last tap registers correctly it is happy. Or if that's not the behavior we would want to test then we should probably remove the test case

**Review by aryamanshodhan — APPROVED** [review]
Well done Johnny! This is very comprehensive.

**Review by ThomasT0830 — APPROVED** [review]
